### ISSUE TITLE
Regression on 6.1.0 for private constructors

### ISSUE THREAD

**Issue by tseho** [phpspec/phpspec#1284]
When upgrading from 6.0.0 to 6.1.0, one of our test is now failing when using a private constructor.

Given a class like this:
```
class Option
{
    private $value;

    private function __construct(string $value)
    {
        $this->value = $value;
    }

    public static function create(string $value)
    {
        return new self($value);
    }
}
```

And the related spec:
```
class OptionSpec extends ObjectBehavior
{
    public function it_is_initializable()
    {
        $this->shouldHaveType(Option::class);
    }

    public function it_can_be_constructed_through_create()
    {
        $this->beConstructedThrough('create', ['foo']);
        $this->shouldBeAnInstanceOf(Option::class);
    }
}
```

We now have, in 6.1.0, the error `exception [exc:ReflectionException("Access to non-public constructor of class Option")] has been thrown`.

I don't understand what changed but I created a repo for reproducing the problem:
with phpspec 6.0.0 : https://github.com/tseho/phpspec-regression-6.1.0/tree/6.0.0
with phpspec 6.1.0 : https://github.com/tseho/phpspec-regression-6.1.0/tree/6.1.0

EDIT:
I overlooked the fact we don't specify the factory in the `it_is_initializable`, so our test was bad but it was passing somehow.

**Comment by ciaranmcnulty** [comment]
This is still an issue IMO if it's breaking behaviour in a minor

**Comment by ciaranmcnulty** [comment]
It's caused by https://github.com/phpspec/phpspec/pull/1267/commits/33efa6a641c52a327906064014fab70673c406c9

Previously `shouldHaveType` wasn't triggering instantiation (which is weird behaviour, see #1233)

I'm up for changing this behaviour for 7.0.0 but we should:
a) add an acceptance test for this case
b) revert the behaviour

I could revert all of PR #1267 for now, unless @ddziaduch wants to take a look?

**Comment by tseho** [comment]
Ok, I haven't been far enough and I didn't know it was previously using `newInstanceWithoutConstructor`.
I agree this change should be included in a major.

**Comment by ddziaduch** [comment]
> I overlooked the fact we don't specify the factory in the it_is_initializable, so our test was bad but it was passing somehow.

Ah yes, when I was working on the fix I got the same thing. The PhpSpec itself was silencing this error. Seems that my fix fix that so your specs are now failing 🙈

@ciaranmcnulty I think we should revert that and put that into major version (7?)

**Comment by tkotosz** [comment]
note: it seems `shouldImplement` has the same problem, now it is trying instantiate the object and fails if the object has a private constructor.

**Comment by ciaranmcnulty** [comment]
Fixed in 6.1.1
